# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1535429068 14400
#      Tue Aug 28 00:04:28 2018 -0400
# Node ID d9d192928b45b38321e8d6602289b045d8917c68
# Parent  2c42ed37cfc27fe8d7509462dc140d69b3bd650d
process tests in @CLASS directories again (bug #54561)

* __run_test_suite__.m: Process tests in @CLASS directories but don't
cd to them as for other directories containing tests.

diff --git a/scripts/testfun/__run_test_suite__.m b/scripts/testfun/__run_test_suite__.m
--- a/scripts/testfun/__run_test_suite__.m
+++ b/scripts/testfun/__run_test_suite__.m
@@ -151,43 +151,45 @@ function [pass, fail, xfail, xbug, skip,
   endif
 
   function [dp, dn, dxf, dxb, dsk, drtsk, drgrs] = run_test_dir (fid, d)
 
     lst = dir (d);
     dp = dn = dxf = dxb = dsk = drtsk = drgrs = 0;
     for i = 1:length (lst)
       nm = lst(i).name;
-      if (lst(i).isdir && nm(1) != "." && ! strcmp (nm, "private")
-          && nm(1) != "@")
+      if (lst(i).isdir && nm(1) != "." && ! strcmp (nm, "private"))
         [p, n, xf, xb, sk, rtsk, rgrs] = run_test_dir (fid, [d, filesep, nm]);
         dp += p;
         dn += n;
         dxf += xf;
         dxb += xb;
         dsk += sk;
         drtsk += rtsk;
         drgrs += rgrs;
       endif
     endfor
 
     saved_dir = pwd ();
     unwind_protect
-      cd (d);
+      [dnm, fnm] = fileparts (d);
+      if (fnm(1) != "@")
+        cd (d);
+      endif
       for i = 1:length (lst)
         nm = lst(i).name;
         ## Ignore hidden files
         if (nm(1) == '.')
           continue
         endif
         if ((length (nm) > 2 && strcmpi (nm((end-1):end), ".m"))
             || (length (nm) > 4
                 && (strcmpi (nm((end-3):end), "-tst")
                     || strcmpi (nm((end-3):end), ".tst"))))
-          p = n = xf = xb = sk = rtsk = 0;
+          p = n = xf = xb = sk = rtsk = rgrs = 0;
           ffnm = fullfile (d, nm);
           ## Only run if contains %!test, %!assert, %!error, %!fail, or %!warning
           if (has_tests (ffnm))
             tmp = reduce_test_file_name (ffnm, topbuilddir, topsrcdir);
             print_test_file_name (tmp);
             [p, n, xf, xb, sk, rtsk, rgrs] = test (ffnm, "quiet", fid);
             print_pass_fail (p, n, xf, xb, sk, rtsk, rgrs);
             dp += p;
